third-party/fb-mysql/8.0.20/man/perror.1
'\" t
.\"     Title: \fBperror\fR
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 03/06/2020
.\"    Manual: MySQL Database System
.\"    Source: MySQL 8.0
.\"  Language: English
.\"
.TH "\FBPERROR\FR" "1" "03/06/2020" "MySQL 8\&.0" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
perror \- display MySQL error message information
.SH "SYNOPSIS"
.HP \w'\fBperror\ [\fR\fB\fIoptions\fR\fR\fB]\ \fR\fB\fIerrorcode\fR\fR\fB\ \&.\&.\&.\fR\ 'u
\fBperror [\fR\fB\fIoptions\fR\fR\fB] \fR\fB\fIerrorcode\fR\fR\fB \&.\&.\&.\fR
.SH "DESCRIPTION"
.PP
\fBperror\fR
displays the error message for MySQL or operating system error codes\&. Invoke
\fBperror\fR
like this:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBperror [\fR\fB\fIoptions\fR\fR\fB] \fR\fB\fIerrorcode\fR\fR\fB \&.\&.\&.\fR
.fi
.if n \{\
.RE
.\}
.PP
\fBperror\fR
attempts to be flexible in understanding its arguments\&. For example, for the
ER_WRONG_VALUE_FOR_VAR
error,
\fBperror\fR
understands any of these arguments:
\fB1231\fR,
001231,
MY\-1231, or
MY\-001231, or
ER_WRONG_VALUE_FOR_VAR\&.
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBperror 1231\fR
MySQL error code MY\-001231 (ER_WRONG_VALUE_FOR_VAR): Variable \*(Aq%\-\&.64s\*(Aq
can\*(Aqt be set to the value of \*(Aq%\-\&.200s\*(Aq
.fi
.if n \{\
.RE
.\}
.PP
If an error number is in the range where MySQL and operating system errors overlap,
\fBperror\fR
displays both error messages:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBperror 1 13\fR
OS error code   1:  Operation not permitted
MySQL error code MY\-000001: Can\*(Aqt create/write to file \*(Aq%s\*(Aq (OS errno %d \- %s)
OS error code  13:  Permission denied
MySQL error code MY\-000013: Can\*(Aqt get stat of \*(Aq%s\*(Aq (OS errno %d \- %s)
.fi
.if n \{\
.RE
.\}
.PP
To obtain the error message for a MySQL Cluster error code, use the
\fBndb_perror\fR
utility\&.
.PP
The meaning of system error messages may be dependent on your operating system\&. A given error code may mean different things on different operating systems\&.
.PP
\fBperror\fR
supports the following options\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-help\fR,
\fB\-\-info\fR,
\fB\-I\fR,
\fB\-?\fR
.sp
Display a help message and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-ndb\fR
.sp
Print the error message for a MySQL Cluster error code\&.
.sp
This option was removed in MySQL 8\&.0\&.13\&. Use the
\fBndb_perror\fR
utility instead\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-silent\fR,
\fB\-s\fR
.sp
Silent mode\&. Print only the error message\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-verbose\fR,
\fB\-v\fR
.sp
Verbose mode\&. Print error code and message\&. This is the default behavior\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-version\fR,
\fB\-V\fR
.sp
Display version information and exit\&.
.RE
.SH "COPYRIGHT"
.br
.PP
Copyright \(co 1997, 2020, Oracle and/or its affiliates. All rights reserved.
.PP
This documentation is free software; you can redistribute it and/or modify it only under the terms of the GNU General Public License as published by the Free Software Foundation; version 2 of the License.
.PP
This documentation is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with the program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA or see http://www.gnu.org/licenses/.
.sp
.SH "SEE ALSO"
For more information, please refer to the MySQL Reference Manual,
which may already be installed locally and which is also available
online at http://dev.mysql.com/doc/.
.SH AUTHOR
Oracle Corporation (http://dev.mysql.com/).


NEWS
"Rogers" 10-Sep-2015
- Update to jemalloc 4.0.0.
- Convert more architecture dependent code to VASM.
- Large bout of fixes for compiling hhvm under MSVC.
- Have hh_emitter support more of HHBC.
- Remove MInstrState from CGetM and IssetM's final operations.
- Refactor [un]serialize functions.
- Use profiling counters to decide whether to apply LICM.
- Move inlining to translateRegion.
- Implement intdiv function.
- Remove joomla and joomlacms from hhvm test suite.
- [hack] Add a --stats option to the client.
- [hack] GC the shared memory area more aggressively.
- [hack] Rework the representation of strings.
- Fix bug #70014 - use RAND_bytes instead of deprecated RAND_pseudo_bytes
- Fix bug #69545 - avoid overflow when reading list and improve fix for bug #69545
- Fix #66387: Stack overflow with imagefilltoborder
- Fix #70064: imagescale(..., IMG_BICUBIC) leaks memory and Fix #69024: imagescale segfault with palette based image

"Quake" 27-Aug-2015
- ??

"Park" 13-Aug-2015
- ???

"Osborne" 30-Jul-2015
- Remove global LDAP requirement
- Don't require libXSLT globally
- Remove the global MCrypt requirement
- Remove the global search for LibUODBC
- Remove the global check for libpam
- use folly::writeFull wrapper for uploads
- Fix vararg handling for array_uintersect and others
- fix error message for abstract type constants
- Fix variadics in hhvm_get_frame_args
- Use poll() instead of select() in XDebugServer::createSocket.
- [hack] Error on overriding concrete type constant with an abstract one
- [hack] Grow shapes more correctly
- Array pointer modification functions (next, current, each etc) can be used on ArrayObject
- Eliminate ext_mbstring's dependence on ext_process
- Eliminate ext_mail's dependence on ext_process
- Use utf8 string literals in zend-html
- Don't allow abstract memoized methods
- [hack] Properly handle 'this' type for methods
- Fix lexing of "${fn()}"
- Fix lexing of #<?php
- [hack] Fix infinite loop in syntax error when closing a namespace
- Fix mispelling of COOKIE.
- Deal with self, static and parent constructors

"Nash" 16-Jul-2015
- Make ArrayData ref-count start at 1
- Remove the NOUSER define when compiling with MSVC, Cygwin, or MinGW
- Fix typos of `__MSC_VER` to be the correct value, `_MSC_VER`
- Shapes::removeKey(...) - remember the names of unset fields
- Shapes::toArray() - add .hhi
- Shapes::removeKey()
- Shapes::keyExists()
- Shapes::idx helper function for accessing optional fields
- Don't assume a scalar Map/Set initializer is an integer if its not a string
- Update mysqlclient library to 5.6.23 everywhere
- Port fix for php bug #48147
- Fixed #5475: ResourceBundle::getLocales('') don't return list of locales
- [hhvm][oss][frameworks] Add Smarty to Framework test runner
- Add in a facility to pool curl handles between requests.
- Fix xdebug trying to write to a nullptr std::FILE*.
- Fix bug when appending unaligned StringSlice
- Remove XDebug chrome support
- Incorrect return value from eval, Closure generated in first eval pass is returned in the second eval pass
- Stop throwing ints in xdebug
- CVE-2015-4663: Support verifying certificates on ssl/tls streams, enable by default
- Allow inlining immutable methods in non-repo mode
- Reusable TC
- Release some old hack design documentation
- Don't step into builtin functions in the debugger
- Support collections in array_map.hhas
- [hhvm] Make sure iteration of class_iterator actually stops
- [hhvm] Open source idx
- Add .hhas implementation of array_reduce
- Change memory layout for hni

"McQueen" 2-Jul-2015
- Add basic stream_socket_enable_crypto() client support
- [hack] Check types for class vars in decl mode as well
- [hhvm] Fix xdebug with multiple commands (and hopefully with PhpStorm)
- [hack][hhvm] Standard library for shapes - runtime support
- Fix date_create_immutable error state
- [hack] Include debug symbols in OSS build
- Update Hack Tools README to point to new OPAM Quick Install URL
- [hack] add color mode to Tty
- Update CMake/FindTBB.cmake for current Windows support.
- [hhvm][mac] Fix ext_ipc on Mac
- [hhvm][easy] Add README.md to php-octane
- [zoncolan] Missing cases to bring BW analysis at par with FW
- [zoncolan] Added "support" for reflection
- [zoncolan] Added support for XHP
- [hack][shapes] Anonymous shapes
- [hack][shapes] Structural subtyping for shapes proposal
- Fix some bugs in ext_servicerouter.cpp / ext_gatehouse.cpp
- [zoncolan] Added analysis of toplevel functions
- [zonconal] Adding shapes
- [hack] Infer types for constants in decl mode as well
- [hhvm] Enable JIT by default on Mac
- [hack] Format: Improve handling of comments in switch statements
- Make ObjectData ref-counts start at 1
- APC collection optimizations
- [Hack] Fix JSON escaping
- [hack] Make Typing_instantiate only work on decl ty
- use ...$args in invariant_violation()
- [hack] Make decl mode equivalent to partial + // UNSAFE
- [hhvm] Don't use libbfd in OSS build by default
- [hhvm][mac] Fix async mysql time calculation on Mac clang

"Lukather" 18-Jun-2015
- Check that the right operand of instanceof is not an enum
- Allow inlining static builtin methods
- Fix hhbbc not emitting RATs for public static properties.
- Fix overflow in array_slice
- Better error message for double quotes used in shape field names
- infer return type of floor/ceil
- Do not translate functions from units marked as interpret-only.
- Eliminate ldimm saveflags operand
- Allow folding the ideomatic hack callable wrappers
- Optimize NativeImpl
- Modify ReleaseVVOrExit to deal with VarEnvs
- Allow inlined 86ctors
- Fix folly build with Clang
- Re-fix region-prune-arcs assertion
- [mac] Fix crash compiling crc32 with Mac clang
- [mac] Use -force_load, not -all_load
- Don't constrain key or value for SetWithRefElem
- Remove IRBuilder::prepareForNextHHBC---just use exceptionStackBoundary
- Use HHBBC types to avoid generic IncRefs in Await translations
- Don't check for failed waithandles until after it wasn't finished

third-party/fb-mysql/8.0.20/utilities/perror.cc
/*
   Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is also distributed with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have included with MySQL.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */

/* Return error-text for system error messages and handler messages */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>

#include "m_ctype.h"
#include "m_string.h"
#include "my_getopt.h"
#include "my_sys.h"
#include "mysys_err.h"
#include "print_version.h"
#include "welcome_copyright_notice.h" /* ORACLE_WELCOME_COPYRIGHT_NOTICE */

static bool verbose;

#include "my_base.h"
#include "my_compiler.h"
#include "mysys/my_handler_errors.h"

static struct my_option my_long_options[] = {
    {"help", '?', "Displays this help and exits.", nullptr, nullptr, nullptr,
     GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"info", 'I', "Synonym for --help.", nullptr, nullptr, nullptr, GET_NO_ARG,
     NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"silent", 's', "Only print the error message.", nullptr, nullptr, nullptr,
     GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"verbose", 'v', "Print error code and message (default).", &verbose,
     &verbose, nullptr, GET_BOOL, NO_ARG, 1, 0, 0, nullptr, 0, nullptr},
    {"version", 'V', "Displays version information and exits.", nullptr,
     nullptr, nullptr, GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {nullptr, 0, nullptr, nullptr, nullptr, nullptr, GET_NO_ARG, NO_ARG, 0, 0,
     0, nullptr, 0, nullptr}};

static void usage(void) {
  print_version();
  puts(ORACLE_WELCOME_COPYRIGHT_NOTICE("2000"));
  printf(
      "Print a description for a system error code or a MySQL error code.\n");
  printf(
      "If you want to get the error for a negative error code, you should "
      "use\n-- before the first error code to tell perror that there was no "
      "more options.\n\n");
  printf("Usage: %s [OPTIONS] [ERRORCODE [ERRORCODE...]]\n", my_progname);
  my_print_help(my_long_options);
  my_print_variables(my_long_options);
}

static bool get_one_option(int optid,
                           const struct my_option *opt MY_ATTRIBUTE((unused)),
                           char *argument MY_ATTRIBUTE((unused))) {
  switch (optid) {
    case 's':
      verbose = false;
      break;
    case 'V':
      print_version();
      exit(0);
    case 'I':
    case '?':
      usage();
      exit(0);
  }
  return false;
}

static int get_options(int *argc, char ***argv) {
  int ho_error;

  if ((ho_error = handle_options(argc, argv, my_long_options, get_one_option)))
    exit(ho_error);

  if (!*argc) {
    usage();
    return 1;
  }
  return 0;
} /* get_options */

static const char *get_ha_error_msg(int code) {
  /*
    If you got compilation error here, check that every HA_ERR_xxx constant
    has a corresponding error message in handler_error_messages[] list
    (check mysys/my_handler_errors.h and include/my_base.h).
  */
  static_assert(
      HA_ERR_FIRST + array_elements(handler_error_messages) == HA_ERR_LAST + 1,
      "Wrong number of elements in handler_error_messages.");
  if (code >= HA_ERR_FIRST && code <= HA_ERR_LAST)
    return handler_error_messages[code - HA_ERR_FIRST];

  return NullS;
}

typedef struct {
  const char *name;
  uint code;
  const char *text;
  /* SQLSTATE */
  const char *odbc_state;
  const char *jdbc_state;
  uint error_index;
} st_error;

static st_error global_error_names[] = {
#include <mysqld_ername.h>

    {nullptr, 0, nullptr, nullptr, nullptr, 0}};

/**
  Lookup an error by code in the global_error_names array.
  @param code the code to lookup
  @param [out] name_ptr the error name, when found
  @param [out] msg_ptr the error text, when found
  @return 1 when found, otherwise 0
*/
int get_ER_error_msg_by_code(uint code, const char **name_ptr,
                             const char **msg_ptr) {
  st_error *tmp_error;

  /* handle "global errors" */
  if ((code >= EE_ERROR_FIRST) && (code <= EE_ERROR_LAST)) {
    *name_ptr = nullptr;
    *msg_ptr = EE(code);
    return 1;
  }

  tmp_error = &global_error_names[0];

  while (tmp_error->name != nullptr) {
    if (tmp_error->code == code) {
      *name_ptr = tmp_error->name;
      *msg_ptr = tmp_error->text;
      return 1;
    }
    tmp_error++;
  }

  return 0;
}

/**
  Lookup an error by symbol in the global_error_names array.
  @param symbol the symbol to lookup
  @return code >0 when found, otherwise 0
*/
int get_ER_error_msg_by_symbol(const char *symbol) {
  st_error *tmp_error = &global_error_names[0];

  while (tmp_error->name != nullptr) {
    if (0 == strcmp(tmp_error->name, symbol)) return tmp_error->code;
    tmp_error++;
  }

  return 0;
}

#if defined(_WIN32)
static bool print_win_error_msg(DWORD error, bool verbose) {
  LPTSTR s;
  if (FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM,
                    NULL, error, 0, (LPTSTR)&s, 0, NULL)) {
    if (verbose)
      printf("Win32 error code %d: %s", error, s);
    else
      puts(s);
    LocalFree(s);
    return 0;
  }
  return 1;
}
#endif

static const char *get_handler_error_message(int nr) {
  return handler_error_messages[nr - HA_ERR_FIRST];
}

/*
  Register handler error messages for usage with my_error()

  NOTES
    This is safe to call multiple times as my_error_register()
    will ignore calls to register already registered error numbers.
*/

void my_handler_error_register() {
  /*
    If you got compilation error here, check that every HA_ERR_xxx constant has
    a corresponding error message in handler_error_messages[] list
    (check mysys/ma_handler_errors.h and include/my_base.h).
  */
  static_assert(
      HA_ERR_FIRST + array_elements(handler_error_messages) == HA_ERR_LAST + 1,
      "Wrong number of elements in handler_error_messages.");
  my_error_register(get_handler_error_message, HA_ERR_FIRST,
                    HA_ERR_FIRST + array_elements(handler_error_messages) - 1);
}

void my_handler_error_unregister(void) {
  my_error_unregister(
      HA_ERR_FIRST, HA_ERR_FIRST + array_elements(handler_error_messages) - 1);
}

int main(int argc, char *argv[]) {
  int error, code, found;
  const char *msg;
  const char *name;
  char *unknown_error = nullptr;
#if defined(_WIN32)
  bool skip_win_message = 0;
#endif
  MY_INIT(argv[0]);

  if (get_options(&argc, &argv)) exit(1);

  my_handler_error_register();

  error = 0;
  {
    /*
      On some system, like Linux, strerror(unknown_error) returns a
      string 'Unknown Error'.  To avoid printing it we try to find the
      error string by asking for an impossible big error message.

      On Solaris 2.8 it might return NULL
    */
    if ((msg = strerror(10000)) == nullptr) msg = "Unknown Error";

    /*
      Allocate a buffer for unknown_error since strerror always returns
      the same pointer on some platforms such as Windows
    */
    unknown_error = static_cast<char *>(malloc(strlen(msg) + 1));
    my_stpcpy(unknown_error, msg);

    for (; argc-- > 0; argv++) {
      found = 0;

      code = atoi(*argv);

      if ((*argv != nullptr) && (strlen(*argv) > 3)) {
        if (0 == strncmp(*argv, "MY-", 3))
          code = atoi(((char *)*argv) + 3);
        else if (0 == strncmp(*argv, "ER_", 3))
          code = get_ER_error_msg_by_symbol((char *)*argv);
      }

      msg = strerror(code);

      /*
        We don't print the OS error message if it is the same as the
        unknown_error message we retrieved above, or it starts with
        'Unknown Error' (without regard to case).
      */
      if (msg &&
          my_strnncoll(&my_charset_latin1, (const uchar *)msg, 13,
                       (const uchar *)"Unknown Error", 13) &&
          (!unknown_error || strcmp(msg, unknown_error))) {
        found = 1;
        if (verbose)
          printf("OS error code %3d:  %s\n", code, msg);
        else
          puts(msg);
      }
      if ((msg = get_ha_error_msg(code))) {
        found = 1;
        if (verbose)
          printf("MySQL error code MY-%06d (handler): %s\n", code, msg);
        else
          puts(msg);
      }
      if (get_ER_error_msg_by_code(code, &name, &msg)) {
        found = 1;
        if (verbose) {
          if (name != nullptr)
            printf("MySQL error code MY-%06d (%s): %s\n", code, name, msg);
          else
            printf("MySQL error code MY-%06d: %s\n", code, msg);
        } else
          puts(msg);
      }
      if (!found) {
#if defined(_WIN32)
        if (!(skip_win_message = !print_win_error_msg((DWORD)code, verbose))) {
#endif
          fprintf(stderr, "Illegal error code: %d\n", code);
          error = 1;
#if defined(_WIN32)
        }
#endif
      }
#if defined(_WIN32)
      if (!skip_win_message) print_win_error_msg((DWORD)code, verbose);
#endif
    }
  }

  /* if we allocated a buffer for unknown_error, free it now */
  if (unknown_error) free(unknown_error);

  exit(error);
}
